### PR TITLE
Release duplicates based on score.

### PR DESCRIPTION
**Pull request by xezero**
Built upon @beeedy 's initial work, reworked the release duplicates logic to be based on the Pokemon score - which is the combined normalized weighted values of CP and IV:

SCORE = SCORE_CP_WEIGHT * (CP / MAX_CP ) + SCORE_IV_WEIGHT * (IV / 100)

By default, the weights are 0.5 for both CP and IV, meaning we want our score to be based equally on how optimal each of those values are.

If we only care about high CP, then we can set our SCORE_CP_WEIGHT to 1.0 and SCORE_IV_WEIGHT to 0, and vice versa.

RELEASE_DUPLICATES_SIMILARITY_RATIO (renamed from RELEASE_DUPLICATES_SCALER) determines how close the Pokemon's score has to be to the current highest Pokemon's score in order to not release it.

**UPDATE:**

**I've refactored it a bit to allow for custom score expressions (default is still the expression above). This also eliminated the need to have SCORE_CP_WEIGHT and SCORE_IV_WEIGHT as configurable options since the whole expression can be defined in SCORE_EXPRESSION now. Much simpler!**

### CODE REVIEW

**Comment by beeedy** [comment]
I don't understand why this involved a changing of logic. Can you explain why you removed the option to pick how lvl was chosen and forced everyone to use this method? You should instead add an option to define lvl as "score" and simply add this as an option in get_pokemon_lvl

**Comment by xezero** [comment]
Well I removed the option because setting the CP or IV weight essentially eliminated the need to set LVL = CP or LVL = IV.

**Comment by beeedy** [comment]
But what if I want to experiment with other options for sorting? You have removed a completely flexible system that needed a couple lines added to define "score" and instead forced this method onto people. I just want to understand your reasoning

**Comment by xezero** [comment]
LVL=CP+IV really makes no sense since you're adding a very large number with a very small number, so your level is still largely determined by CP. Same with LVL=CP*IV. The numbers make more sense when you normalize the individual values, which is essentially what the score is. It's basically what LVL is, but normalized.

**Comment by xezero** [comment]
You have a point with adding more flexibility. I only removed the other lvl options because they were handled by score. I can re-add the ability to define score in a custom way if you want. Let me make a quick change then.

**Comment by xezero** [comment]
@beeedy Added the ability to evaluate custom score expressions.

File: pgoapi/pgoapi.py [hunk 1]
```diff
@@ -519,6 +524,19 @@ def get_caught_pokemons(self, inventory_items):
...
+                    MAX_CP = pokemon.max_cp
+                    IV = pokemon.iv
+                    try:
+                        pokemon.score = eval(self.CUSTOM_SCORE_EXPRESSION)
+                    except Exception as e:
+                        self.log.error("Error in evaluating custom score expression %s", e)
+                else:
+                    if pokemon.max_cp > 0 and pokemon.iv > 0:
+                        pokemon.score = pokemon.iv / 100.0 * self.SCORE_IV_WEIGHT + pokemon.cp / pokemon.max_cp * self.SCORE_CP_WEIGHT
+
```
Comment (destiny117) [inline pgoapi/pgoapi.py]: You should not calculate this here, instead additional pokemon attributes should be calculated directly inside the `Pokemon()` class within pokemon.py

**Comment by destiny117** [comment]
I noticed in your pokemon list in #171 that your score calculation is not really accurate, it more resemples the IV than anything else

furthermore your code still removes the @beeedy s functionality and replaces it with yours

**Comment by xezero** [comment]
@destiny117 Which part of it is inaccurate? I'll do one by hand:

`2016-07-27 18:48:54,854 [    pgoapi] [ INFO]  Type: Omanyte, CP: 567, IV: 64.4444444444, Max CP: 607.866120538, Score: 0.788607813426
 `

567/607.866*0.5 + 64.44/100.0*0.5 = 0.789

**Comment by xezero** [comment]
And I've removed some functionality since it's redundant and still added a way to make custom expressions per @beeedy's comment. And I don't see anything wrong with modifying functionality so long as the same end goal is achieved, so not sure what you're getting at.

File: pgoapi/pgoapi.py (hunk 1 above)
Comment (xezero) [inline pgoapi/pgoapi.py]: I would indeed like to do that. Would you help me figure out the best way to pass those config constants into pokemon.py? I'm not a Python convention expert by any means, but I felt like passing constants through the Pokemon initializer seemed like a bad idea.

**Comment by destiny117** [comment]
if you look at the first two in your list:
```
Type: Vaporeon, CP: 1942, IV: 82.2222222222, Max CP: 1971.41902867, Score: 0.903649727144
Type: Vaporeon, CP: 1346, IV: 86.6666666667, Max CP: 1367.93131356, Score: 0.925317099864
```
you either have not considered their CP value correctly within the formular or something else is wrong, caus in my opinion the 1942 CP Vaporeon should definitly have a higher score

File: pgoapi/pgoapi.py (hunk 1 above)
Comment (xezero) [inline pgoapi/pgoapi.py]: @destiny117 Check out my refactor per your comments! I've eliminated some of the redundant configuration options and now just simply use **SCORE_EXPRESSION** and passed that through to the Pokemon class initializer to calculate the score inside there. Thanks for the suggestion! :)

**Comment by destiny117** [comment]
looking at the code i found the issue:
`self.max_cp = (attack * sqrt(defense) * sqrt(stamina) * cp_multiplier * cp_multiplier) / 10`
cp_multiplier is actually a representation of the pokemon level, if you want to have actual max_cp you have to make the formula as follows:
`self.max_cp = (attack * sqrt(defense) * sqrt(stamina) * pow(0.79030001,2)) / 10`
because according to http://pokemongo.gamepress.gg/cp-multiplier cp multipliers increase with each level and currenlty max out at 0.79030001 on level 40

**Comment by destiny117** [comment]
Also since the official CP calculation already considers IV values as part of its calculation, an actual score formula along the following lines might be more accurate:
```
current_max_level = 40
self.level = get_level_by_cpm(self.cp_multiplier + self.additional_cp_multiplier) # to be implemented
self.score = self.iv / 100.0 * 0.5 + 100 / current_max_level * self.level * 0.5
```

or even more accurate by considering the trainer-level:
```
max_level = trainer_level + 1.5
self.level = get_level_by_cpm(self.cp_multiplier + self.additional_cp_multiplier) # to be implemented
self.score = self.iv / 100.0 * 0.5 + 100 / max_level * self.level * 0.5
```

**Comment by xezero** [comment]
Very nice find! Yeah I was hesitant to divide the CP by the absolute max CP since that would give a pretty low CP subscore if the trainer was lower level. I thought the cp_multiplier to be used in the equation would just be simply self.cp_multiplier + self.additional_cp_multiplier based on:

`CPM = Base_CPM + Additional_CPM`

from http://pokemongo.gamepress.gg/pokemon-stats-advanced, but am I misreading something?

**Comment by destiny117** [comment]
no that is correct as far as i can tell, since Base_CPM is the pokemons wild level, Additional_CP is the leveles gained by powerups, combining them gives you the current CPM (= current level)

**Comment by xezero** [comment]
Ah, looks like the problem here is additional_cp_multiplier is returning 0 for everything :(

Since I've substituted the corresponding cp_multiplier from http://pokemongo.gamepress.gg/cp-multiplier for my trainer level and the computed max_cp values make more sense.

So I think I need to either figure out how to fix additional_cp_multiplier or just hard-code the cp_multiplier from that table and use the one for the trainer level.

**Comment by destiny117** [comment]
additional_cp_multiplier is correctly 0 for most pokemon, because you probably did not power them up manually. it will only be set once you power them up.

check the formula i provided above which converts CPM to a level first (has to be implemented with the cp->level table first), it should be more accurate than using CP or CPM since CPM is not increasing linear

**Comment by xezero** [comment]
Should be fixed now :)

**Comment by destiny117** [comment]
i did not test it but i expect it will give much more accurate results now already?

things i still don't like about the PR:
- I dont think CP_Multipliers fits inside the Player-class, rather in the Pokemon class. and it is actually CP_Multipliers_Max_Per_level
- I dont think it's a good idea to let people define `SCORE_EXPRESSION` which is being evaluated during runtime, i prefered @beeedy s method more which just gives a few choices for sorting (that you remove in this PR)
- cp multiplier can actually be a bit higher than what you have, since you can boost a pokemon 1.5 levels over your current level
- your formula still compares CP and IV, even though CP is a calculated value that already includes IV

for the last 2 reasons i believe it would be better to calculate and use the pokemon level instead of CP (i'm preparing a formula atm)

as you know i'm working on my own implementation atm so i might be biased, let's see what others think

**Comment by beeedy** [comment]
> I dont think it's a good idea to let people define SCORE_EXPRESSION which is being evaluated during runtime

Agree 100%. If you accidentally have an error in your expression you can accidentally release a lot of Pokemon. I still have not seen a reason that can't just be added into my solution, as my solution was designed to be flexible for the very reason of allowing people to add their own definitions for lvl. I can't look at two Pokemon and instantly tell which the algorithm will pick, and that bothers me.

**Comment by xezero** [comment]
I can resort back to having to define SCORE_CP_WEIGHT and have the CUSTOM_SCORE_EXPRESSION only be active if USE_CUSTOM_EXPRESSION is enabled. By doing so, the user is at their own risk, just like modifying the code directly - the risk of accidentally releasing a lot of Pokemon is about the same.

I can move the constants to the Pokemon class if you think it's a more suitable locaion - I figured since it was referring to the Max_CP_Multiplier specificly tied to a trainer's level, that it made more sense in the Player class, but the decision is subjective, I don't particularly care where it goes lol

The idea was to derive at a very good default score expression that would work for 90% or more of users, whatever that score expression ends up being, whether the formula as is now, or what @destiny117 comes up with. So far the formula I have now seems to be an improvement over the previous ones, but I have no hesitation in adopting a better one if one arises...

In terms of the formula itself, it's not comparing CP to IV directly, it's comparing the ratio of CP to MaxCP (which is independent of IV - this value is dependent on strictly trainer level) and combining that with the ratio of IV to MaxIV (which is a constant 100), so the individual ratios are  independent of each other. Boosting CP with candy makes the Pokemon's CP stronger which makes the score higher, as expected?!

**Comment by xezero** [comment]
Updated the code with the latest changes from develop, which includes @destiny117's new maxCP calculations.

Here's the scores using my current formula but with the new maxCP values:

```
2016-07-28 12:16:09,264 [    pgoapi] [ INFO] Type: Omanyte CP: 790, IV: 71.11, Lvl: 26.0, LvlWild: 26.0, MaxCP: 950, MaxCPAbs: 1064, IV-Norm.: 77, Score: 0.77
2016-07-28 12:16:09,264 [    pgoapi] [ INFO] Type: Omanyte CP: 201, IV: 64.44, Lvl: 7.0, LvlWild: 7.0, MaxCP: 923, MaxCPAbs: 1034, IV-Norm.: 63, Score: 0.43
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Omanyte CP: 650, IV: 64.44, Lvl: 22.0, LvlWild: 22.0, MaxCP: 924, MaxCPAbs: 1035, IV-Norm.: 65, Score: 0.67
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Omanyte CP: 764, IV: 71.11, Lvl: 26.0, LvlWild: 26.0, MaxCP: 919, MaxCPAbs: 1030, IV-Norm.: 63, Score: 0.77
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Omastar CP: 1655, IV: 73.33, Lvl: 27.0, LvlWild: 27.0, MaxCP: 1916, MaxCPAbs: 2146, IV-Norm.: 74, Score: 0.80
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Omastar CP: 1679, IV: 80.00, Lvl: 27.0, LvlWild: 27.0, MaxCP: 1944, MaxCPAbs: 2177, IV-Norm.: 83, Score: 0.83
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Omastar CP: 1460, IV: 97.78, Lvl: 23.0, LvlWild: 23.0, MaxCP: 1984, MaxCPAbs: 2222, IV-Norm.: 97, Score: 0.86
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabuto CP: 828, IV: 66.67, Lvl: 28.0, LvlWild: 28.0, MaxCP: 925, MaxCPAbs: 1036, IV-Norm.: 72, Score: 0.78
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabuto CP: 874, IV: 82.22, Lvl: 29.0, LvlWild: 29.0, MaxCP: 943, MaxCPAbs: 1056, IV-Norm.: 80, Score: 0.87
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabuto CP: 469, IV: 60.00, Lvl: 16.0, LvlWild: 16.0, MaxCP: 917, MaxCPAbs: 1027, IV-Norm.: 69, Score: 0.56
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabuto CP: 874, IV: 75.56, Lvl: 30.0, LvlWild: 30.0, MaxCP: 911, MaxCPAbs: 1020, IV-Norm.: 66, Score: 0.86
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabuto CP: 698, IV: 80.00, Lvl: 23.0, LvlWild: 23.0, MaxCP: 949, MaxCPAbs: 1062, IV-Norm.: 83, Score: 0.77
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabuto CP: 931, IV: 88.89, Lvl: 30.0, LvlWild: 30.0, MaxCP: 971, MaxCPAbs: 1087, IV-Norm.: 93, Score: 0.92
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabutops CP: 1532, IV: 77.78, Lvl: 26.0, LvlWild: 26.0, MaxCP: 1842, MaxCPAbs: 2063, IV-Norm.: 80, Score: 0.80
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabutops CP: 1794, IV: 84.44, Lvl: 30.0, LvlWild: 30.0, MaxCP: 1869, MaxCPAbs: 2093, IV-Norm.: 89, Score: 0.90
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Kabutops CP: 1685, IV: 91.11, Lvl: 28.0, LvlWild: 28.0, MaxCP: 1881, MaxCPAbs: 2107, IV-Norm.: 93, Score: 0.90
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Aerodactyl CP: 1266, IV: 80.00, Lvl: 21.0, LvlWild: 21.0, MaxCP: 1884, MaxCPAbs: 2110, IV-Norm.: 83, Score: 0.74
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Snorlax CP: 2029, IV: 62.22, Lvl: 24.0, LvlWild: 23.0, MaxCP: 2643, MaxCPAbs: 2960, IV-Norm.: 63, Score: 0.69
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Dratini CP: 394, IV: 66.67, Lvl: 15.0, LvlWild: 15.0, MaxCP: 823, MaxCPAbs: 921, IV-Norm.: 71, Score: 0.57
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Dratini CP: 374, IV: 77.78, Lvl: 14.0, LvlWild: 14.0, MaxCP: 836, MaxCPAbs: 936, IV-Norm.: 78, Score: 0.61
2016-07-28 12:16:09,265 [    pgoapi] [ INFO] Type: Dragonair CP: 613, IV: 64.44, Lvl: 13.0, LvlWild: 13.0, MaxCP: 1475, MaxCPAbs: 1652, IV-Norm.: 68, Score: 0.53
2016-07-28 12:16:09,266 [    pgoapi] [ INFO] Type: Dragonite CP: 1480, IV: 86.67, Lvl: 15.0, LvlWild: 15.0, MaxCP: 3084, MaxCPAbs: 3453, IV-Norm.: 89, Score: 0.67
2016-07-28 12:16:09,266 [    pgoapi] [ INFO] Type: Dragonite CP: 3046, IV: 88.89, Lvl: 32.0, LvlWild: 13.0, MaxCP: 3071, MaxCPAbs: 3440, IV-Norm.: 86, Score: 0.94
```

How do these scores look?

**Comment by xezero** [comment]
One interesting analysis here for:

```
Type: Kabuto CP: 874, IV: 75.56, Lvl: 30.0, LvlWild: 30.0, MaxCP: 911, MaxCPAbs: 1020, IV-Norm.: 66, Score: 0.86
Type: Kabuto CP: 698, IV: 80.00, Lvl: 23.0, LvlWild: 23.0, MaxCP: 949, MaxCPAbs: 1062, IV-Norm.: 83, Score: 0.77
```

is even though the second Kabuto has a higher IV-Norm and IV, it's score is still lower than the first Kabuto since the first one has a higher CP to MaxCP ratio (another note is that we can see their respective MaxCP's are almost the same, so the first Kabuto with it's already higher CP makes it more attractive to keep than the second one).

**Comment by destiny117** [comment]
please have i look at PR #216 where i reimplemented #110 into the new realease algorythm.
specifically you should check out the function pokemon_score and the corresponding config sections.
I put a new method "FANCY" in there which tries to acomplish your goal.

It's using the pokemon level relative to player level instead of the CP_Max value, because really i don't think you should compare those values since CP always incorporates the IV already and you could just sort by IV

**Comment by xezero** [comment]
Lvl does seem to work too, but doesn't Lvl have a positive correlation with CP anyways? If you took a look at the Kabuto example above, the second Kabuto has both a higher IV-Norm and IV, even though the first Kabuto would be better to pick since it has an already higher CP (and therefore higher score) with an IV that's not much lower.

**Comment by destiny117** [comment]
yes it has, CP is a combination of level, base stats (always the same) and IV :P that's why i would either just use CP or compare IV directly to the level

**Comment by xezero** [comment]
Makes sense :)

**Pull request closed without merging**